src/Akka.Persistence.Sql.Linq2Db/Journal/DAO/BaseByteArrayJournalDao.cs

```diff
@@ -200,45 +195,61 @@ await db.GetTable<JournalRow>()
+        //And are able to have a little more control over default capacity of array.
+        static List<JournalRow> FlattenListOfListsToList(List<Akka.Util.Try<List<JournalRow>>> source) {
             
-            /*var trySet = new List<JournalRow>();
-            foreach (var serializedTry in serializedTries)
+            //List<JournalRow> ResultSet(
+            //    Akka.Util.Try<List<JournalRow>> item)
+            //{
+            //    return item.Success.GetOrElse(new List<JournalRow>(0));
+            //}
+
+            List<JournalRow> rows = new List<JournalRow>(source.Count > 4 ? source.Count:4);
+            for (var index = 0; index < source.Count; index++)
             {
-                trySet.AddRange(serializedTry.Success.GetOrElse(new List<JournalRow>(0)));
+                var item = source[index].Success.Value;
+                if (item != null)
+                {
+                    rows.AddRange(item);
+                }
+                //rows.AddRange(ResultSet(source[index]));
             }
 
-            var rows = Seq(trySet);*/
-            var rows = Seq(serializedTries.SelectMany(serializedTry =>
-                    serializedTry.Success.GetOrElse(new List<JournalRow>(0)))
-                .ToList());
-            //
+            return rows;
+        }
         
+        public async Task<IImmutableList<Exception>> AsyncWriteMessages(
+            IEnumerable<AtomicWrite> messages, long timeStamp = 0)
+        {
+            var serializedTries = Serializer.Serialize(messages, timeStamp);
             
-            return await QueueWriteJournalRows(rows).ContinueWith(task =>
-                {
-                    //We actually are trying to interleave our tasks here...
-                    //Basically, if serialization failed our task will likely
-                    //Show success
-                    //But we instead should display the serialization failure
-                    return serializedTries.Select(r =>
-                        r.IsSuccess
-                            ? (task.IsFaulted
-                                ? TryUnwrapException(task.Exception)
-                                : null)
-                            : r.Failure.Value).ToImmutableList();
-                }, CancellationToken.None,
-                TaskContinuationOptions.ExecuteSynchronously,
-                TaskScheduler.Default);
+            //Fold our List of Lists into a single sequence
+            var rows = Seq(FlattenListOfListsToList(serializedTries));
+            //Wait for the write to go through. If Task fails, write will be captured
+            //As WriteMessagesFailure.
+            await QueueWriteJournalRows(rows);
+            //If we get here, we build an ImmutableList containing our rejections.
+            //These will be captured as WriteMessagesRejected
+            return BaseByteArrayJournalDao
+                .BuildWriteRejections(serializedTries);
```

Comment:
We were doing the wrong thing before, Plz doublecheck against [original scala](https://github.com/akka/akka-persistence-jdbc/blob/63238b87f26b60af5c72494211a5f8943ef6ff66/core/src/main/scala/akka/persistence/jdbc/journal/dao/legacy/ByteArrayJournalDao.scala#L73-L86).

The idea here -should- be that if QueueWriteJournalRows fails, or the write fails, we throw here and generate Failure. Otherwise, we will see if any of the writes failed serialization and treat those as Rejected.

Reply:
Makes sense to me - the rows that go through should be a success but the ones that encountered failures should get a rejected message